File: dsa-problems/SegmentTree/solutions/python/SEG-012-range-add-kth-order.py
import sys
from bisect import bisect_right

def process(arr: list[int], ops: list[list[str]]) -> list[int]:
    n = len(arr)
    # Heuristic block size
    block_size = int((n * 10) ** 0.5) # Slightly larger block size
    if block_size < 100: block_size = 100
    
    blocks = []
    # Each block: [sorted_vals, lazy]
    # We also need original arr updated lazily?
    
    # Initialize blocks
    for i in range(0, n, block_size):
        chunk = arr[i : i + block_size]
        blocks.append([sorted(chunk), 0])
        
    results = []
    
    def update(l, r, x):
        start_block = l // block_size
        end_block = r // block_size
        
        if start_block == end_block:
            # Partial
            b = blocks[start_block]
            # Push lazy to arr
            if b[1] != 0:
                for i in range(start_block * block_size, min(n, (start_block + 1) * block_size)):
                    arr[i] += b[1]
                b[1] = 0
            
            # Update arr
            for i in range(l, r + 1):
                arr[i] += x
                
            # Rebuild sorted
            chunk = arr[start_block * block_size : min(n, (start_block + 1) * block_size)]
            b[0] = sorted(chunk)
            
        else:
            # Start partial
            b = blocks[start_block]
            if b[1] != 0:
                for i in range(start_block * block_size, (start_block + 1) * block_size):
                    arr[i] += b[1]
                b[1] = 0
            for i in range(l, (start_block + 1) * block_size):
                arr[i] += x
            b[0] = sorted(arr[start_block * block_size : (start_block + 1) * block_size])
            
            # Middle full
            for i in range(start_block + 1, end_block):
                blocks[i][1] += x
                
            # End partial
            b = blocks[end_block]
            if b[1] != 0:
                for i in range(end_block * block_size, min(n, (end_block + 1) * block_size)):
                    arr[i] += b[1]
                b[1] = 0
            for i in range(end_block * block_size, r + 1):
                arr[i] += x
            b[0] = sorted(arr[end_block * block_size : min(n, (end_block + 1) * block_size)])

    def count_le(l, r, val):
        count = 0
        start_block = l // block_size
        end_block = r // block_size
        
        if start_block == end_block:
            lazy = blocks[start_block][1]
            for i in range(l, r + 1):
                if arr[i] + lazy <= val:
                    count += 1
        else:
            lazy = blocks[start_block][1]
            for i in range(l, (start_block + 1) * block_size):
                if arr[i] + lazy <= val:
                    count += 1
            
            for i in range(start_block + 1, end_block):
                b = blocks[i]
                target = val - b[1]
                count += bisect_right(b[0], target)
                
            lazy = blocks[end_block][1]
            for i in range(end_block * block_size, r + 1):
                if arr[i] + lazy <= val:
                    count += 1
        return count

    for op in ops:
        if op[0] == "ADD":
            update(int(op[1]), int(op[2]), int(op[3]))
        else:
            l, r, k = int(op[1]), int(op[2]), int(op[3])
            low = -2 * 10**14
            high = 2 * 10**14
            ans = high
            
            while low <= high:
                mid = (low + high) // 2
                if count_le(l, r, mid) >= k:
                    ans = mid
                    high = mid - 1
                else:
                    low = mid + 1
            results.append(ans)
            
    return results

def main():
    import sys
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    it = iter(input_data)
    n = int(next(it))
    q = int(next(it))
    arr = [int(next(it)) for _ in range(n)]
    ops = []
    for _ in range(q):
        type = next(it)
        if type == "ADD":
            ops.append([type, next(it), next(it), next(it)])
        else:
            ops.append([type, next(it), next(it), next(it)])
    
    results = process(arr, ops)
    for res in results:
        print(res)

if __name__ == "__main__":
    main()

File: dsa-problems/SegmentTree/solutions/javascript/SEG-012-range-add-kth-order.js
class Solution {
  process(arr, ops) {
    // Use BigInt for values to be safe, though 2e14 fits in double (safe integer limit 9e15).
    // Standard numbers are fine.
    
    const n = arr.length;
    // Heuristic block size
    let blockSize = Math.floor(Math.sqrt(n * Math.log2(n + 1))) + 1;
    if (blockSize < 100) blockSize = 100;
    
    const numBlocks = Math.ceil(n / blockSize);
    const blocks = [];
    
    // Initialize blocks
    for (let i = 0; i < numBlocks; i++) {
      const start = i * blockSize;
      const end = Math.min(n, start + blockSize);
      const sorted = arr.slice(start, end).sort((a, b) => a - b);
      blocks.push({ sorted: new Float64Array(sorted), lazy: 0 });
    }
    
    // We need a mutable array for partial updates
    // arr is already mutable
    
    const results = [];
    
    const upperBound = (arr, val) => {
      let l = 0, r = arr.length;
      while (l < r) {
        const mid = (l + r) >>> 1;
        if (arr[mid] > val) r = mid;
        else l = mid + 1;
      }
      return l;
    };

    const countLessEqual = (l, r, val) => {
      let count = 0;
      const startBlock = Math.floor(l / blockSize);
      const endBlock = Math.floor(r / blockSize);
      
      if (startBlock === endBlock) {
        const lazy = blocks[startBlock].lazy;
        for (let i = l; i <= r; i++) {
          if (arr[i] + lazy <= val) count++;
        }
      } else {
        const lazyStart = blocks[startBlock].lazy;
        const startLimit = (startBlock + 1) * blockSize;
        for (let i = l; i < startLimit; i++) {
          if (arr[i] + lazyStart <= val) count++;
        }
        
        for (let i = startBlock + 1; i < endBlock; i++) {
          const b = blocks[i];
          const target = val - b.lazy;
          count += upperBound(b.sorted, target);
        }
        
        const lazyEnd = blocks[endBlock].lazy;
        const endStart = endBlock * blockSize;
        for (let i = endStart; i <= r; i++) {
          if (arr[i] + lazyEnd <= val) count++;
        }
      }
      return count;
    };

    const update = (l, r, x) => {
      const startBlock = Math.floor(l / blockSize);
      const endBlock = Math.floor(r / blockSize);
      
      const partialUpdate = (bIdx, l, r) => {
        const b = blocks[bIdx];
        const start = bIdx * blockSize;
        const end = Math.min(n, start + blockSize);
        
        if (b.lazy !== 0) {
          for (let i = start; i < end; i++) arr[i] += b.lazy;
          b.lazy = 0;
        }
        
        for (let i = l; i <= r; i++) arr[i] += x;
        
        // Rebuild sorted
        // Creating new Float64Array is fast enough
        const chunk = arr.slice(start, end).sort((a, b) => a - b);
        b.sorted = new Float64Array(chunk);
      };

      if (startBlock === endBlock) {
        partialUpdate(startBlock, l, r);
      } else {
        partialUpdate(startBlock, l, (startBlock + 1) * blockSize - 1);
        for (let i = startBlock + 1; i < endBlock; i++) {
          blocks[i].lazy += x;
        }
        partialUpdate(endBlock, endBlock * blockSize, r);
      }
    };

    for (const op of ops) {
      if (op[0] === "ADD") {
        update(parseInt(op[1]), parseInt(op[2]), parseInt(op[3]));
      } else {
        const l = parseInt(op[1]);
        const r = parseInt(op[2]);
        const k = parseInt(op[3]);
        
        let low = -200000000000000;
        let high = 200000000000000;
        let ans = high;
        
        while (low <= high) {
          const mid = Math.floor((low + high) / 2);
          if (countLessEqual(l, r, mid) >= k) {
            ans = mid;
            high = mid - 1;
          } else {
            low = mid + 1;
          }
        }
        results.push(ans);
      }
    }
    return results;
  }
}

const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => {
  const parts = line.trim().split(/\s+/).filter(x => x !== "");
  for (const p of parts) data.push(p);
});
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const q = parseInt(data[idx++], 10);
  const arr = [];
  for (let i = 0; i < n; i++) arr.push(parseInt(data[idx++], 10));
  const ops = [];
  for (let i = 0; i < q; i++) {
    const type = data[idx++];
    ops.push([type, data[idx++], data[idx++], data[idx++]]);
  }
  const solution = new Solution();
  const out = solution.process(arr, ops);
  console.log(out.join("\n"));
});
